Core/GDCore/Events/CodeGeneration/EventsCodeGenerationContext.h

```diff
@@ -269,10 +308,14 @@ class GD_CORE_API EventsCodeGenerationContext {
                               ///< inheriting from context with depth n.
   unsigned int
       customConditionDepth;  ///< The depth of the conditions being generated.
+  unsigned int asyncDepth;
   unsigned int* maxDepthLevel;  ///< A pointer to a unsigned int updated with
                                 ///< the maximum depth reached.
   const EventsCodeGenerationContext*
       parent;  ///< The parent of the current context. Can be NULL.
+  std::set<EventsCodeGenerationContext*>
+      asyncRoots;  ///< The first parent of the current and previous async

```

Comment:
I see the idea but had to read the code. Do you think we could avoid storing them at all with some kind of recursive calls to the `parent`?

Reply:
🤔 Maybe, indeed, but I'm afraid that wouldn't be optimized (I think I added a todo to another function to switch from a recursive parent call to using this for that reason too), since not all parents but only the ones with this.asyncDepth != parent.asyncDepth should be affected by this, so if we deal with a lot of subevents and few wait actions, we could be iterating in a lot of contexts that are irrelevant which could slow down the code generation. 
But then again we would also spare the copy of a vector, so not sure what the good move in that regard is.

Reply:
I see. Is there a way in this case to only remember the "closestAsyncParent"? And navigate recursively from closestAsyncParent to closestAsyncParent?

From what I remember/understood this only used to declare to the (async) parents the objects that are declared in their descendance (i.e: all their child contexts).

Reply:
🤔 Sounds like a good idea, I'll try that.

Reply:
Hmmm...
I am probably doing a stupid and obvious mistake, but I have the following code:
```cpp
    if (IsAsync()) {
      gd::EventsCodeGenerationContext* asyncContext = nearestAsyncParent;
      while (asyncContext != nullptr) {
        asyncContext->allListsAcrossChildrenToBeDeclared.insert(objectName);
        if (asyncContext->parent == nullptr ||
            asyncContext->parent->nearestAsyncParent == nullptr)
          break;
        asyncContext = asyncContext->parent->nearestAsyncParent;
      }
    }
```

But WASM crashes when running that, saying that there was an out-of-bounds memory access... I put at every step a check to make sure we were not using a null pointer by accident, and I don't see what else could cause this error...

For context, `InheritsFrom` does `nearestAsyncParent = parent_.nearestAsyncParent;` and `AsyncInheritsFrom` does `nearestAsyncParent = this;`

Reply:
Are you sure you've initialised nearestAsyncParent to `nullptr` in all the possible constructors of EventsCodeGenerationContext? Otherwise it will contain garbage memory, so a pointer to anywhere.
You should be able to debug by `std::cout << ... << std::flush;` the pointers and see if there is anything suspicious. Don't forget the flush otherwise a crash won't have the time to display the log because logged output is cached before being sent to the console.

Reply:
(otherwise push your WIP code on a separate branch to see if I can find something suspicious :))

Reply:
> Are you sure you've initialised nearestAsyncParent to nullptr

Ah, that was it, thanks! I thought `nullptr` was a pointers default value, it works like a charm now :)